### ISSUE TITLE
Meeting notes for GSoC JSON and Excel import project

### ISSUE THREAD

**Issue by dmos62** [centerofci/mathesar#3028]
Each post is meant to contain the notes for one meeting, except the first, which contains the archive of all previous meetings.

Until week #7 we used this HackMD document (private): https://hackmd.io/gXFfAA7JTvGP0LbK8-tH6A.

Related meta tickets:

- https://github.com/centerofci/mathesar/issues/2895
- https://github.com/centerofci/mathesar/issues/3027
- https://github.com/centerofci/mathesar/issues/3026

**Comment by dmos62** [comment]
## 2023-07-07 (week #6)
### participants
- Anshuman
- Anish

### meeting todo
- Discussed about the DataFile class and its implementations for the backend for different file types.

### post-meeting todo
- Anshuman
    - create a proposal about import flow for handling different types of data files in the backend
    - continue working on adding `max_level` to the endpoint.
    - open a PR for copy-paste json.

## 2023-07-05 (week #6) (async)
### participants
- Dom
- Anshuman

I pointed out to Anshuman that we want `max_level` JSON normalization parameter to be an API parameter. Anshuman also found a bug where if we import data that has an `id` column, and its values are not all non-null or non-unique, backend will crash with a 500.

### post-meeting todo
- anshuman
    - Report the bad id bug,
        - and postpone working on it until end of project (if there's time)
    - Add the `max_level` API parameter
        - Make sure docs are updated accordingly

### things to discuss during upcoming meetings
- our todos and plan tracking is scattered
    - we use this meeting note document
        - in this document we use
            - post-meeting todo lists
            - and the "week plans"
    - and the github meta issue (that I (Dom) admittedly had forgotten about)
        - https://github.com/centerofci/mathesar/issues/2895
    - would be nice to fix this

## 2023-07-05 (week #6)
### participants
- Dom
- Anshuman
- Anish

### meeting todo
- extending copy-paste import to accept JSON
- progress discussion
- JSON import documentation

## 2023-06-27 (week #5)
### participants
- Dom
- Anshuman

### meeting todo
- discuss getting JSON import documentation done
- elaborate plan for week 5 and onwards
- check if there's anything to be done for the project midpoint
- elaborate what excel we accept, and what is the ideal excel

### post-meeting todo
- anshuman
    - make sure JSON import docs describe clearly what kinds/forms of JSON we can accept
    - make changes to instructions in import UI
        - regarding what filetypes are accepted

### ideal excel and what excel we'll accept

- import a single sheet only
    - provide API parameter for which sheet (index) to use with the default being first
    - we can have a UI element to select the index
        - though that's outside the scope of this project
- what's the ideal?
    - single sheet
    - single table in that sheet
    - with a header
- what file extension do we expect?
    - whatever pandas accepts?

### plan for the coming weeks
- week 5
    - tasks
        - Merge JSON import documentation
        - Merge corrections to instructions in import UI
- week 6
    - tasks
        - Merge functionality to accept perfect Excel
        - Merge API tests for that
- week 7
    - tasks
        - Merge Excel normalization code and return correct errors from backend.
        - Merge API tests for that
- week 8
    - tasks
        - Merge API parameter for sheet index

## 2023-06-22 (week #4)
### participants
- Dom
- Anshuman

### meeting todo
- discuss last week's async progress report
- are we on track?
    - we're close to the midpoint of the project
- resolve design problem related to json normalization
- get a clear picture of what "API tests" mean
- discuss informing user of what kinds of json are accepted
- discuss displaying nice errors

### anshuman's perspective
- feels on track, expects to get everything done by end of june
    - end of june is project's midpoint
- not worried about JSON normalization
- worried about UI/UX
    - how to inform user of what forms json is accepted in?

### nice errors
- how do we return errors from our service layer
    - example
        - https://github.com/centerofci/mathesar/blob/develop/mathesar/api/serializers/records.py
    - we use custom errors
        - they're defined under `mathesar.api.exceptions.*.*`
        - the example above illustrates how we convert generic errors to them
        - they're often raised in our Django serializers (as in the example)
            - but can be equally raised in any other code within the `mathesar` namespace
- ideally you'll preserve any information we have when the exceptions are thrown
    - meaning make sure that information ends up in our custom exceptions
    - however, if that information is not easily accessible, you can ignore it

### post-meeting todo
- anshuman
    - document a description of what kinds of JSON are accepted
        - involves a PR that updates our documentation
            - our documentation
                - https://github.com/centerofci/mathesar/tree/develop/docs
        - done in a way that it can be linked

## 2023-06-13 (week #3)
### participants
- Dom (absent)
- Anshuman

At Dom's request, Anshuman prepared [an async progress report](https://hackmd.io/64hKdM-RS9mZz4dP0Vntrg). Below is the copy-paste of that report. Normal sync meeting was not conducted.

### copy-paste beginning

### LAST MEETING TODO
- Remove duplicated logic between mathesar/imports/csv.py and json.py files.
    - <span style="color:Green"> Completed </span>
    - PR [#2926](https://github.com/centerofci/mathesar/pull/2926)
- Create a bug report for a JSON object (serialized to string) being castable to URI
    - <span style="color:Green"> Completed </span>
    - Issue [#2941](https://github.com/centerofci/mathesar/issues/2941)
- Postpone handling of json normalization edge cases that are not handled by pandas
    - <span style="color:Green"> Completed </span>
    - Issue [#2943](https://github.com/centerofci/mathesar/issues/2943)


### WEEKLY TASKS
- Merge API tests for everything written until this point
    - <span style="color:Green"> Completed </span>
    - PR [#2933](https://github.com/centerofci/mathesar/pull/2933)
- Merge JSON normalization code and return correct errors from backend.
    - <span style="color:Orange"> In progress </span>
    - Waiting on the PR #2926 to get merged. Changes made in the above PR will affect the file where we call normalization functions.


### Doubts
- I am not getting the clear picture of what "API tests" mean. Could you please point me to these tests in our codebase so I can follow the same procedure for importing JSON too?

### copy-paste end

## 2023-06-06 (week #2)

meeting todo
- discuss PR #2906

PR #2906
- duplicated logic between mathesar/imports/csv.py and json.py files
    - can be solved by creating a common/base file and deduplicating/centralizing some of that logic
    - refactoring/deduplication can be done in a new PR, to make reviewing easier
    - part of the problem is that the methods in `csv.py` file are too big (some of them)
        - splitting them up into smaller methods will likely help code reuse/refactoring
            
enums for enumerating file types
- example of enum use in our codebase
    - https://github.com/centerofci/mathesar/blob/74ebe8d42455b28a08bbb2f705afee7fb7ad206a/db/types/base.py#LL91C12-L91C12
    - you might be able to find simpler uses of enums in our codebase
        - try grepping for `Enum`
- not a huge deal, but enums would improve developer experience somewhat

post-meeting todo
- anshuman
    - create a bug report for a JSON object (serialized to string) being castable to URI
    - postpone handling of json normalization edge cases that are not handled by pandas
        - create a github issue to track them
        - don't forget to mention this issue in your root meta ticket
    - consider using enums for file types

Updated plan for implementation JSON importing
- week 2
    - tasks
        - Merge API tests for everything written until this point
            - milestone reached
                - Getting a perfect JSON accepted
        - Merge JSON normalization code and return correct errors from backend.
- week 3
    - tasks
        - Merge remaining tests
            - milestone reached
                - Conversion and Normalization
- week 4
    - tasks
        - Improve UI for better user understanding of the new feature. Display errors properly.
            - consider meeting with other members of the team
                - to discuss how to communicate what kind of JSON is accepted
            - milestone reached
                - UI and errors

## 2023-05-30 (week #1)

meeting todo
- set objectives for this week
- setup meta ticket

Anshuman's plan for implementation JSON importing
- week 0
    - tasks
        - Update Datafile object (Introduce a 'type' attribute) and store JSON as a datafile
            - done
- week 1
    - tasks
        - Merge essential utility functions to get perfect JSON working
            - get_column_names_from_json()
            - create_db_table_from_json_data_file()
            - create_table_from_json()
        - Update functions that insert data into tables
            - create_table_from_datafile()
            - insert_records_from_json()
- week 2
    - tasks
        - Merge API tests for everything written until this point
            - milestone reached
                - Getting a perfect JSON accepted
        - Merge functions to convert invalid JSON structures to valid one.
- week 3
    - tasks
        - Merge JSON normalization code and return correct errors from backend.
        - Merge remaining tests
            - milestone reached
                - Conversion and Normalization
- week 4
    - tasks
        - Improve UI for better user understanding of the new feature. Display errors properly.
            - consider meeting with other members of the team
                - to discuss how to communicate what kind of JSON is accepted
            - milestone reached
                - UI and errors

after-meeting todo
- anshuman
    - consider creating an issue ticket for fixing datafiles being kept forever
    - add summary to the meta ticket
    
## *Earlier notes redacted*

**Comment by dmos62** [comment]
## 2023-07-11 (week #7)
### participants
- Anshuman
- Dom

### meeting todo
- our todos and plan tracking is scattered
- are we on track?
- splitting DataFile class into multiple classes/subclasses (per type)

### post-meeting todo
- anshuman
  - [ ] create an issue for the frontend to adhere to new importable filetypes
  - [ ] create an issue for the backend to split the DataFile class into multiple per-filetype classes
  - [ ] create the importing excel meta ticket

### plan for the coming weeks
- week 7
  - [ ] Merge `max_level` API param
  - [ ] Merge functionality to accept perfect Excel
- week 8
  - [ ] Merge API tests for importing perfect Excel
  - [ ] Merge Excel normalization code and return correct errors from backend.
  - [ ] Merge API tests for rejecting bad Excel or importing normalized Excel
- week 9
  - [ ] Do postponed/additional tasks

**Comment by Anish9901** [comment]
## 2023-07-14 (week 7)
### participants
- Anshuman
- Anish

### meeting todo
- Why we keep datafiles stored after the import is completed sucessfully?
- Discussion about #3030 & #3031

### post-meeting todo
Anshuman:
- [ ] Finish #3039
- [ ] Merge functionality to accept perfect Excel

Anish:
- Research about #2910

**Comment by dmos62** [comment]
## 2023-07-18 (week 8)
### participants
- Anshuman
- Dom

### meeting todo
- discuss progress
  - previous week's todos
  - are we on track?
- upcoming midterm evaluation
- discuss Excel PR #3059
  - in some cases file name extension is not available, or unknown
    - have to resort to bruteforce checking
### post-meeting todo
Anshuman:
- [ ] explain rationale for choosing algo in docstring of `_get_file_type` 
- [ ] for file type inference, use file name extension when available, otherwise resort to bruteforcing
  - if that seems unpractical, resort to bruteforcing always

**Comment by Anish9901** [comment]
## 2023-07-28 (week 8)
### participants
- Anshuman
- Anish

### meeting todo
- Progress discussion
- What excel normalization would entail?

### post-meeting todo
Anshuman:
- Address requested changes in #3059 
- Start to work on Excel normalization.

**Comment by dmos62** [comment]
## 2023-08-02 (week 10)
### participants
- Anshuman
- Dom

### meeting todo
- Progress discussion
  - Made progress, but PR work is blocked on #3059
  - #3083 and Excel normalization code is close enough to merging to schedule them for this week
- Getting #3059 merged
  - Slight changes left only


### plan for the coming weeks
- week 10
  - [ ] #3059
  - [ ] #3083
  - [ ] Merge Excel normalization code and return correct errors from backend.
- week 11
  - [ ] Merge API tests for rejecting bad Excel or importing normalized Excel
- week 12
  - [ ] Do postponed/additional tasks
- week 20
  - Final deadline (2023-09-25 Monday)

**Comment by Anish9901** [comment]
## 2023-08-04 (week 10)

### participants
- Anshuman
- Anish

### meeting todo
- Progress discussion

### post-meeting todo
Anshuman:
- Functionality for excel normalization & its API tests.

**Comment by dmos62** [comment]
## 2023-08-08 (week 11)
### participants
- Anshuman
- Dom

### meeting todo
- Progress discussion
  - Didn't end up merging #3083 and #3132 during last week
  - Didn't come up with a lot of tests for rejecting bad Excel or for normalization
    - Making that a part of #3132
    - We'll figure out if the tests we have are sufficient during PR reviews
- Start scheduling [postponed/additional tasks
](https://github.com/centerofci/mathesar/issues/3026)
  - Reorder tasks so that higher tasks are higher priority
- Shortening the project to 16 weeks
  - Anshuman noticed that he can complete the project faster than the time we've allotted (estimated as 16 weeks)
    - Dom's thoughts: having those few additional weeks as reserve is nice, would prefer to not surrender them
  - He's also interested in demonstrating his ability to deliver faster, which might be useful when evaluating him for a more permanent position in Mathesar
    - Dom's thoughts: project length will not affect our evaluation of Anshuman's skill or competency.

### post-meeting todo
- Dom
  - Let Kriti know that Anshuman is interested in joining Mathesar on a more permanent basis

### plan for the coming weeks
- week 11
  - [ ] #3083
  - [ ] #3132
  - [ ] #3009
- week 12
  - [ ] #3031
- week 14
  - [ ] #3030
- week 20
  - Final deadline (2023-09-25 Monday)

**Comment by Anish9901** [comment]
## 2023-08-18(week 12)

### participants
- Anshuman
- Anish

### meeting todo
- Progress discussion
    - Anshuman wasn't able to do much progress as he had fallen ill, so the last weeks tasks are caried forward. 
    - Starting to work on #3009

### post-meeting todo
Anshuman:
- [ ] Make #3083 mergable.
- [ ] #3009 

Anish:
- Merge #3132 
There seems to be some problem with github actions that is causing the tests to fail. The PR is ready to merge though.

**Comment by Anish9901** [comment]
## 2023-08-25(week 13)

### participants
- Anshuman
- Anish

### meeting todo
- Progress discussion
    - Anshuman wasn't able to make much progress last wee due to his acads which will be compensated over the coming weekend.
    - [ ] Fix errors in #3083 & #3132 & merge them.

### post-meeting todo
Anshuman:
- [ ] Add tests to #3155

**Comment by Anish9901** [comment]
## 2023-09-01(week 14)

### participants
- Anshuman
- Anish

### meeting todo
- Progress discussion
    - Reviewed #3195, Waiting for a review from @dmos62. 
    - Will start to work on #3031, after approval from the frontend team as the complexity seems higher than initially expected.  

### post-meeting todo
Anshuman:
- [ ] Make a PR for #3031

**Comment by dmos62** [comment]
## 2023-09-05(week 17)

### participants
- Anshuman
- Anish

### meeting todo
- Progress discussion
  - Three full weeks left until final deadline
  - Two ["postponed tasks"](https://github.com/centerofci/mathesar/issues/3026) left
    - That's excluding the blocked "allow different JSON structures" issue
      - Blocked, because we decided that it shouldn't be done unless users request it
- Discuss #3195
  - Nothing of note, wil re-review when Anshuman's made requested changes
- Anshuman will be doing a lightining talk about his GSoC project and Mathesar
  - On September 20th

### plan for the coming weeks
- week 15
  - [ ] Merge #3195
  - [ ] Raise PR for #3031
- week 16
  - [ ] Merge PR for #3031
  - [ ] Raise PR for #3030
- week 17
  - [ ] Merge PR for #3030
  - [ ] Raise PR for documentation
- week 18
  - [ ] Final deadline (2023-09-25 Monday)
  - [ ] Anshuman's doing a lightning talk with GSoC (2023-09-28 Thursday)

**Comment by Anish9901** [comment]
## 2023-09-08(week 15)

### participants
- Anshuman
- Anish

Same as https://github.com/centerofci/mathesar/issues/3028#issuecomment-1702749482

**Comment by dmos62** [comment]
## 2023-09-12 (week 16)

### participants
- Anshuman
- Dom

### meeting todo
- Progress discussion
  - Didn't have time for #3031
  - Submitted documentation PR for review #3204 (should be merged after #3031)
  - Feels on track to get remaining work done by the deadline (in the next two weeks)
    - Last exam is on the 15th September

**Comment by Anish9901** [comment]
## 2023-09-15 (week 16)

### participants
- Anshuman
- Anish

### meeting todo
- Progress discussion
    - Anshuman wasn't able to make progress this week due to his exams, but will cover the backlog in the coming week.

**Comment by dmos62** [comment]
## 2023-09-19 (week 17)

### participants
- Anshuman
- Dom

### meeting todo
- Progress discussion
- Final report
  - Dom is ok with it being hosted in our wiki or in our issue tracker
  - Anshuman is interested in using a meta-ticket for that

### post-meeting todo
- Anshuman
    - [ ] ask Sean for guidance or help in preparing a UX design  
    - [ ] if a meta-ticket is to be used for the final report, a new one should be created, because currently we don't have a meta-ticket that encompasses all of our sub-meta-tickets; except, meeting notes, but those aren't great for the final report
    - [ ] some of these meta-tickets have links to the wiki that are broken: fix them
